test/useLifecycleInterfaceRule.spec.ts

```diff
@@ -0,0 +1,193 @@
+          }
+        `;
+        assertFailures(ruleName, source, [
+          {
+            endPosition: {
+              character: 20,
+              line: 2
+            },
+            message: getFailureMessage(LifecycleInterfaces.OnInit, LifecycleMethods.ngOnInit, className),
+            startPosition: {
+              character: 12,
+              line: 2
+            }
+          },
+          {
+            endPosition: {
+              character: 23,
+              line: 4
+            },
+            message: getFailureMessage(LifecycleInterfaces.OnDestroy, LifecycleMethods.ngOnDestroy, className),
+            startPosition: {
+              character: 12,
+              line: 4
+            }
+          }
+        ]);
+      });
+
+      it('should fail when some of the lifecycle hooks are used without implementing their interfaces', () => {
+        const source = `
+          class ${className} extends Component implements OnInit {
+            ngOnInit() {}
+
+            ngOnDestroy() {
+            ~~~~~~~~~~~
+            }
+          }
+        `;
+        assertAnnotated({
+          message: getFailureMessage(LifecycleInterfaces.OnDestroy, LifecycleMethods.ngOnDestroy, className),
+          ruleName,
+          source
+        });
+      });
+    });
+
+    describe('invalid declaration of lifecycle hooks, using ng.hookName', () => {
+      it('should fail when lifecycle hooks are used without implementing all interfaces, using ng.hookName', () => {
+        const source = `
+          class ${className} extends Component implements ng.OnInit {
+            ngOnInit() {}
+
+            ngOnDestroy() {
+            ~~~~~~~~~~~
+            }
+          }
+        `;
+        assertAnnotated({
+          message: getFailureMessage(LifecycleInterfaces.OnDestroy, LifecycleMethods.ngOnDestroy, className),
+          ruleName,
+          source
+        });
+      });
+    });
+  });
+
+  describe('success', () => {
+    describe('valid declaration of lifecycle hooks', () => {
+      it('should succeed when lifecycle hook is used with its corresponding interface', () => {
+        const source = `
+          class ${className} implements OnInit {
+            ngOnInit() {}
+          }
+        `;
+        assertSuccess(ruleName, source);
+      });
+
+      it('should succeed when lifecycle hooks are used with their corresponding interfaces', () => {
+        const source = `
+          class ${className} extends Component implements OnInit, OnDestroy  {
+            ngOnInit() {}
+
+            private ngOnChanges = '';
+
+            ngOnDestroy() {}
+
+            ngOnSmth {}
+          }
+        `;
+        assertSuccess(ruleName, source);
+      });
+    });
+
+    describe('valid declaration of lifecycle hooks, using ng.hookName', () => {
+      it('should succeed when lifecycle hook is used with its interface', () => {
+        const source = `
+          class ${className} implements ng.OnInit {
```

Comment:
Question: using namespace is a good practice ?

Reply:
I really don't know, but I never saw someone using this in real world.

Reply:
@mgechev , what do you think about it ?

Reply:
This may have some implications in the dead code elimination, but it's fine to exist in our test cases.